src/layers/conv.jl

```diff
@@ -925,3 +927,86 @@ function LSTMAggregator(args...; kwargs...)
     in_ch = args[1][1]
     return SAGEConv(args..., Flux.LSTMCell(in_ch, in_ch); kwargs...)
 end
+
+"""
+    EGNNConv((in_dim, int_dim, out_dim); init)
+"""
+
+struct EGNNConv <: MessagePassing
+   nn_edge
+   nn_x
+   nn_h
+
+   in_dim
+   int_dim
+   out_dim
+end
+
+@functor EGNNConv
+
+function EGNNConv(dims::NTuple{3,Int}; init=glorot_uniform)
+    in_dim, int_dim, out_dim = dims
+
+    m_len = in_dim * 2 + 2
+
+    nn_edge = Flux.Dense(m_len, int_dim; init=init)
+
+    nn_x = Flux.Dense(int_dim, 1; init=init)
+    nn_h = Flux.Dense(in_dim + int_dim, out_dim; init=init)
+
+    return EGNNConv(nn_edge, nn_x, nn_h, dims...)
+end
+
+function EGNNConv(nn_edge, nn_x, nn_h; init=glorot_uniform)
+
+    # Assume that these are strictly MLPs (no conv)
+    nn_edge.init(init)
+    nn_x.init(init)
+    nn_h.init(init)
+
+    in_dim = nn_edge.layers[1].W |> x->size(x)[2]
+    int_dim = nn_edge.layers[end].W |> x->size(x)[1]
+    out_dim = nn_h.layers[end].W |> x->size(x)[1]
+    return EGNNConv(nn_edge, nn_x, nn_h, in_dim, int_dim, out_dim)
+end
+
+function message(egnn::EGNNConv, v_i, v_j, e)
+    in_dim = egnn.in_dim
+    h_i = v_i[1:in_dim]
+    h_j = v_j[1:in_dim]
+
+    x_i = v_i[in_dim+1:end]
+    x_j = v_j[in_dim+1:end]
+
+    if isnothing(e)
+        a = 1
+    else
+        a = e[1]
+    end
+
+    input = vcat(h_i, h_j, norm(x_i - x_j, 2)^2, a)
+    edge_msg = egnn.nn_edge(input)
```

Comment:
I think here we can have some API design. Since `EGNN` is a general network, user could define their own `nn_edge`, `nn_x` and `nn_h`. We could provide them as interfaces for user to overwrite on. For example:

- `ϕ_edge`: in terms of `nn_edge`, which accepts `(h_i, h_j, dist, a)` as arguments.
- `ϕ_x`: in terms of `nn_x`, which accepts `(m_ij, )` as arguments.
- `update`: it is itself `update` function in terms of `nn_h`, which accepts `(h_i, m_i)` as arguments.

`h_i`, `h_j`, `norm(x_i - x_j, 2)^2`, `a` should not be `vcat` together, since they are all arguments to `nn_edge`.

Reply:
I was planning to include both paths, but I could see how continuing to include this path would be confusing.

The code I write does a lot of dimensionality checks (in here, and in the `update` function). It would be ideal if

* The user can supply their own arbitrary operators, and
* Upon construction of the layer, the dimensionalities are checked and the resulting `in_dim`, `int_dim`, and `out_dim` are all stored as the layer's fields.

I know that this can be done if we assume that the input functions are, for example, `Flux.Chain` objects consisting only `Flux.Dense` layers. Is there a way that this can be done for completely arbitrary input functions?

Reply:
OK, I will demonstrate this for you.